plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -450,57 +467,141 @@ export class PersonState {
+                        this.teamId,
+                        distinctId1,
+                        tx
+                    )
 
-        //  The last case: (!oldPerson && !newPerson)
-        return [
-            await this.createPerson(
-                // TODO: in this case we could skip the properties updates later
-                timestamp,
-                this.eventProperties['$set'] || {},
-                this.eventProperties['$set_once'] || {},
-                teamId,
-                null,
-                true,
-                this.event.uuid,
-                [mergeIntoDistinctId, otherPersonDistinctId],
-                addDistinctIdVersion
-            ),
-            Promise.resolve(),
-        ]
+                    // See comment above about `distinctIdVersion`
+                    const insertedDistinctId2 = await this.db.addPersonlessDistinctIdForMerge(
+                        this.teamId,
+                        distinctId2,
+                        tx
+                    )
+
+                    // `createPerson` uses the first Distinct ID provided to generate the Person
+                    // UUID. That means the first Distinct ID definitely doesn't need an override,
+                    // and can always use version 0. Below, we exhaust all of the options to decide
+                    // whether we can optimize away an override by doing a swap, or whether we
+                    // need to actually write an override. (But mostly we're being verbose for
+                    // documentation purposes)
+                    let distinctId2Version = 1 // TODO: Once `posthog_personlessdistinctid` is backfilled, this should be = 0
+                    if (insertedDistinctId1 && insertedDistinctId2) {
+                        // We were the first to insert both (neither was used for Personless), so we
+                        // can use either as the primary Person UUID and create no overrides.
+                    } else if (insertedDistinctId1 && !insertedDistinctId2) {
+                        // We created 1, but 2 was already used for Personless. Let's swap so
+                        // that 2 can be the primary Person UUID and no override is needed.
+                        ;[distinctId1, distinctId2] = [distinctId2, distinctId1]
+                    } else if (!insertedDistinctId1 && insertedDistinctId2) {
+                        // We created 2, but 1 was already used for Personless, so we want to
+                        // use 1 as the primary Person UUID so that no override is needed.
+                    } else if (!insertedDistinctId1 && !insertedDistinctId2) {
+                        // Both were used in Personless mode, so there is no more-correct choice of
+                        // primary Person UUID to make here, and we need to drop an override by
+                        // using version = 1 for Distinct ID 2.
+                        distinctId2Version = 1
+                    }
```

Comment:
Covering all the branches like this is really helpful for thinking through this, thanks for writing it all out.